cilium: Fix parsing of embedded JSON

jrajahalme opened the pull request:
'cilium endpoint get' returns embedded JSON (JSON within JSON as a
quoted string). The parsing of the returned embedded JSON has broken
down; this commit fixes that.

The problem was with the definition of the regular expression used to
match the embedded JSON from the unquoted string. By the spec of the
Perl RE '.' matches any character BUT the newline, so we need to
explicitly allow newlines with '(.|\n)'. When newlines are properly
matched by the regular expression, the alternative Match() on the
quoted string is no longer necessary. Besides, the alternative Match()
only worked if the unquoted string did not contain any newlines, and
contained only the JSON, because the match was successful on any part
of the quited string containing the embedded JSON, not only when the
whole byte array was JSON.

Finally, FindAllStringIndex() returns successive, non-overlapping
matches, so taking the last only works when there is onlly one
match. In this case the simpler FindStringIndex(), which returns the
left-most match works equally well.

The unit tests are also enhanced:
- make them run in non-privileged mode
- verify results
- add an example of a 'cilium endpoint get' output, which failed before
  this commit

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8267)
<!-- Reviewable:end -->

joestringer commented on cilium/cmd/helpers_test.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-// +build privileged_tests
+// +build !privileged_tests
```
Unfortunately I believe that this is privileged because the endpoint package directly depends on BPF bits, which depend on the bootstrapping of the BPF filesystem, which causes this test to fail in unprivileged mode on Travis. More invasive refactoring would be required to loosen this requirement.

aanm replied:
We can still have 2 files, one that runs privileged tests and other that doesn't run privileged tests.

jrajahalme replied:
The problem here is that the imports need the bpf mount, even if the test would not use it. The dev VM has that mount so it worked for me locally, but it failed for this reason in Travis. I reverted this change.

joestringer reviewed (approved):
Apart from the likely cause of the failure in Travis, LGTM. Glad to get better test coverage here too.

jrajahalme committed:
cilium: Fix parsing of embedded JSON

'cilium endpoint get' returns embedded JSON (JSON within JSON as a
quoted string). The parsing of the returned embedded JSON has broken
down; this commit fixes that.

The problem was with the definition of the regular expression used to
match the embedded JSON from the unquoted string. By the spec of the
Perl RE '.' matches any character BUT the newline, so we need to
explicitly allow newlines with '(.|\n)'. When newlines are properly
matched by the regular expression, the alternative Match() on the
quoted string is no longer necessary. Besides, the alternative Match()
only worked if the unquoted string did not contain any newlines, and
contained only the JSON, because the match was successful on any part
of the quited string containing the embedded JSON, not only when the
whole byte array was JSON.

Finally, FindAllStringIndex() returns successive, non-overlapping
matches, so taking the last only works when there is onlly one
match. In this case the simpler FindStringIndex(), which returns the
left-most match works equally well.

The unit tests are also enhanced:
- make them run in non-privileged mode
- verify results
- add an example of a 'cilium endpoint get' output, which failed before
  this commit

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
